funsor/joint.py

```diff
@@ -244,6 +245,11 @@ def eager_independent(joint, reals_var, bint_var):
     return None  # defer to default implementation
 
 
+@substitute.register(Joint, dict)
+def substitute_joint(expr, subs):
+    return expr.eager_subs(tuple(subs.items()))
```

Comment:
This is a bit of a hack, but we should be able to get rid of it since `Joint` does not introduce fresh variables.

Reply:
Can you explain why this is a hack?

Reply:
Semantically, `substitute_joint` should not need to do anything `Joint`-specific and should be able to reuse the generic `substitute_funsor` logic from `terms.py`, but that doesn't quite work yet so I'm just calling `Joint.eager_subs`.

Reply:
I still don't understand why "that doesn't quite work yet", so it's probably worth adding a comment like `#TODO remove this unnecessary method after ___`

Reply:
> I still don't understand why "that doesn't quite work yet"

If I comment those lines out then a bunch of tests in `test_joint.py` fail due to missing patterns for `Delta` and `Joint`.  It's a more complicated version of the case of `Affine` where I had to add a pattern so that the result of a generic substitution could be normalized by the interpreter.